mergekit/measure.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (C) 2024 Charles O. Goddard
+#
+# This software is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This software is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see http://www.gnu.org/licenses/.
+
+import logging
+
+import tqdm
+import transformers
+
+from mergekit.architecture import get_architecture_info
+from mergekit.config import MergeConfiguration
+from mergekit.graph import Executor
+from mergekit.io.tasks import LoaderCache
+from mergekit.options import MergeOptions
+from mergekit.plan import MergePlanner
+from mergekit.merge import _model_out_config
+
+
+def run_measure(
+    merge_config: MergeConfiguration,
+    out_path: str,
+    options: MergeOptions,
+):
+    if options.random_seed is not None:
+        transformers.trainer_utils.set_seed(options.random_seed)
+
+    if not merge_config.models and not merge_config.slices:
+        raise RuntimeError("No output requested")
+
+    model_arch_info = [
+        get_architecture_info(m.config(trust_remote_code=options.trust_remote_code))
+        for m in merge_config.referenced_models()
+    ]
+    if not options.allow_crimes:
+        if not all(a == model_arch_info[0] for a in model_arch_info[1:]):
+            raise RuntimeError(
+                "Must specify --allow-crimes to attempt to mix different architectures"
+            )
+    arch_info = model_arch_info[0]
+
+    # initialize loader cache and set options
+    loader_cache = LoaderCache()
+    loader_cache.setup(options=options)
+
+    # create config for output model
+    cfg_out = _model_out_config(
+        merge_config, arch_info, trust_remote_code=options.trust_remote_code
+    )
+
+    # warm up loader cache
+    for model in (
+        pbar := tqdm.tqdm(
+            merge_config.referenced_models(),
+            desc="Warmup loader cache",
+            disable=options.quiet,
+        )
+    ):
+        loader_cache.get(model)
+    del pbar
+
+    logging.info("Planning operations")
+    targets = MergePlanner(
+        merge_config,
+        arch_info,
+        options=options,
+        out_model_config=cfg_out,
+    ).plan_to_disk(out_path=out_path)
+
+    exec = Executor(
+        tasks=targets,
+        math_device="cuda" if options.cuda else "cpu",
+        storage_device="cuda" if options.low_cpu_memory else "cpu",
+    )
+
+    res = []
+    for _task, value in exec.run(quiet=options.quiet):
```

Comment:
Looking this over, I kinda think we might not need a separate file here - maybe it should just early out in `merge.py` if there's a `metric_method` set instead of `merge_method?`